File: ruby/palindrome_linked_list.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end
#
# @param {ListNode} head
# @return {Boolean}
def is_palindrome(head)
  return true if head.nil? or head.next.nil?

  slow, fast = head, head
  until fast.next.nil? or fast.next.next.nil?
    slow, fast = slow.next, fast.next.next
  end

  pre = nil
  while slow
    slow.next, slow, pre = pre, slow.next, slow
  end

  first, second = head, pre
  while first and second
    unless first.val == second.val
      return false
    end
    first, second = first.next, second.next
  end
  return true
end

File: python/palindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return True

        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        pre = None
        while slow:
            slow.next, slow, pre = pre, slow.next, slow
        first, second = head, pre

        while first and second:
            if first.val != second.val:
                return False
            first, second = first.next, second.next
        return True
